### PR TITLE
DAOS-9953: Usability improvements to dmg pool list and query.

### PR DESCRIPTION
**Pull request by ravalsam**
- Adding CLI option to filter rebuild status!= idle from dmg pool list command.
- Show pool state as Degraded when any targets are missing or device are evicted.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/proto/mgmt/pool.proto [hunk 1]
```diff
@@ -145,6 +145,7 @@ message ListPoolsResp {
 		string label = 2; // pool label
 		repeated uint32 svc_reps = 3; // pool service replica ranks
 		string state = 4; // pool state
+		string rebuild_stat = 5; // pool rebuild status
```
Comment (kjacque) [inline src/proto/mgmt/pool.proto:148]: naming - how about `rebuild_state`?

File: src/control/cmd/dmg/pretty/pool.go [hunk 2]
```diff
@@ -303,17 +310,39 @@ func printListPoolsRespVerbose(out io.Writer, resp *control.ListPoolsResp) error
...
+
 	}
 
 	fmt.Fprintln(out, formatter.Format(table))
 
 	return nil
 }
 
+// UpdateListPoolsResponse will update the pool which has been rebuild and not in idle state
+func UpdateListPoolsResponse(final_resp *control.ListPoolsResp, resp *control.ListPoolsResp, rebuildOnly bool) error {
```
Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool.go:322]: style - Go naming convention is camelCase, so should be "finalResp"

Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool.go:322]: I think this helper function should be in the dmg package and unexported. Modifying the response itself feels a bit outside the scope of the pretty package, which is about formatting output text.

File: src/control/cmd/dmg/pretty/pool_test.go [hunk 3]
```diff
@@ -30,14 +30,15 @@ func TestPretty_PrintPoolQueryResp(t *testing.T) {
 		"empty response": {
 			pqr: &control.PoolQueryResp{},
 			expPrintStr: `
-Pool , ntarget=0, disabled=0, leader=0, version=0
+Pool , ntarget=0, disabled=0, leader=0, version=0, state=
```
Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool_test.go:33]: We should try to print something sane even if the state is empty. Maybe "unknown"?

File: src/control/cmd/dmg/pretty/pool_test.go [hunk 4]
```diff
@@ -1197,9 +1203,31 @@ no pools in system
...
+						TargetsTotal:     16,
+						TargetsDisabled:  0,
+						State:            system.PoolServiceStateReady.String(),
+						PoolLayoutVer:    1,
+						UpgradeLayoutVer: 2,
+					},
+				},
+			},
+			verbose: true,
+			expPrintStr: `
```
Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool_test.go:1227]: noQuery is not actually set to true for this test case. Based on the code I would expect the final column to not have been included.

File: src/control/server/mgmt_pool.go [hunk 5]
```diff
@@ -919,6 +920,16 @@ func (svc *mgmtSvc) PoolQuery(ctx context.Context, req *mgmtpb.PoolQueryReq) (*m
...
 
+	ps, err := svc.getPoolService(req.GetId())
+	if err != nil {
+		return nil, err
+	}
+
+	resp.State = ps.State.String()
+	if resp.DisabledTargets > 0 {
+		resp.State = fmt.Sprintf("%s-%s", resp.State, "Degraded")
+	}
```
Comment (kjacque) [inline src/control/server/mgmt_pool.go:931]: This logic should be tested in a unit test.

File: src/control/lib/control/pool.go [hunk 6]
```diff
@@ -1200,6 +1204,7 @@ func ListPools(ctx context.Context, rpcClient UnaryInvoker, req *ListPoolsReq) (
 		p.PoolLayoutVer = resp.PoolLayoutVer
 		p.UpgradeLayoutVer = resp.UpgradeLayoutVer
 		p.setUsage(resp)
+		p.RebuildStat = fmt.Sprintf("%s", resp.Rebuild.State)
```
Comment (kjacque) [inline src/control/lib/control/pool.go:1207]: minor - A simplification suggestion:

```suggestion
		p.RebuildStat = resp.Rebuild.State.String()
```

File: src/control/lib/control/pool.go [hunk 7]
```diff
@@ -1047,6 +1048,9 @@ type (
 
 		// Usage contains pool usage statistics for each storage tier.
 		Usage []*PoolTierUsage `json:"usage"`
+
+		// PoolRebuildStatus contains detailed information about the pool rebuild process.
+		RebuildStat string `json:"rebuild"`
```
Comment (kjacque) [inline src/control/lib/control/pool.go:1053]: minor - Name in the comment needs to match the variable name

File: src/control/cmd/dmg/pretty/pool_test.go [hunk 8]
```diff
@@ -1232,10 +1260,10 @@ Label UUID                                 State SvcReps SCM Size SCM Used SCM I
...
------ ----                                 -----      ------- -------- -------- ------------- --------- --------- -------------- -------- -------------- 
-one   00000001-0001-0001-0001-000000000001 Ready      [0-2]   100 GB   80 GB    12%           6.0 TB    5.0 TB    1%             0/16     1->2           
-two   00000002-0002-0002-0002-000000000002 Destroying [3-5]   100 GB   80 GB    12%           6.0 TB    5.0 TB    1%             8/64     None           
+Label UUID                                 State      SvcReps SCM Size SCM Used SCM Imbalance NVME Size NVME Used NVME Imbalance Disabled UpgradeNeeded? Rebuild State 
+----- ----                                 -----      ------- -------- -------- ------------- --------- --------- -------------- -------- -------------- ------------- 
+one   00000001-0001-0001-0001-000000000001 Ready      [0-2]   100 GB   80 GB    12%           6.0 TB    5.0 TB    1%             0/16     1->2                         
+two   00000002-0002-0002-0002-000000000002 Destroying [3-5]   100 GB   80 GB    12%           6.0 TB    5.0 TB    1%             8/64     None
```
Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool_test.go:1266]: This test should include something in the Rebuild State to show it is printed.

File: src/control/cmd/dmg/pretty/pool.go [hunk 9]
```diff
@@ -303,17 +310,39 @@ func printListPoolsRespVerbose(out io.Writer, resp *control.ListPoolsResp) error
...
+	// Update the pool state to Degraded if any pool target is disabled.
+	for _, pool := range resp.Pools {
+		if pool.TargetsDisabled > 0 {
+			pool.State = fmt.Sprintf("%s-%s", pool.State, "Degraded")
+		}
```
Comment (kjacque) [inline src/control/cmd/dmg/pretty/pool.go:327]: This is already done in the server-side changes, so should be removed here. I think server side is the right place for it.

**Review by kjacque — CHANGES_REQUESTED** [review]

**Comment by knard-intel** [comment]
@ravalsam the functional tests errors are known issue with the pragma `Feature: control`.
https://daosio.atlassian.net/browse/DAOS-13886

A workaround to this issue could be to use the following pragma instead of the  `Feature: control` one.
`Test-tag: pr control,-version_number,-test_daos_control_config_basic,-test_daos_agent_config_basic`

**Review by knard-intel — APPROVED** [review]
LGTM

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

File: src/control/lib/control/pool.go [hunk 10]
```diff
@@ -610,6 +611,24 @@ func PoolQuery(ctx context.Context, rpcClient UnaryInvoker, req *PoolQueryReq) (
 	return pqr, convertMSResponse(ur, pqr)
 }
 
+func UpdatePoolQueryState(pqr *PoolQueryResp) {
+	// Pool state is unknown as default, if TotalTargets is 0.
+	if pqr.TotalTargets == 0 {
+		pqr.State = "Unknown"
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/pool.go:617]: also you could probably return here

File: src/control/cmd/dmg/pretty/pool_test.go [hunk 11]
```diff
@@ -1197,6 +1204,29 @@ no pools in system
 			},
 			verbose: true,
 			expPrintStr: `
+Label UUID                                 State SvcReps SCM Size SCM Used SCM Imbalance NVME Size NVME Used NVME Imbalance Disabled UpgradeNeeded? Rebuild State 
+----- ----                                 ----- ------- -------- -------- ------------- --------- --------- -------------- -------- -------------- ------------- 
+-     00000001-0001-0001-0001-000000000001 Ready N/A     100 GB   80 GB    12%           6.0 TB    5.0 TB    1%             0/16     1->2
```
Comment (tanabarr) [inline src/control/cmd/dmg/pretty/pool_test.go:1209]: why is there no text for rebuild state? we should probably print something

File: src/control/lib/control/pool_test.go [hunk 12]
```diff
@@ -507,6 +507,68 @@ func TestControl_PoolCreate(t *testing.T) {
...
+					DisabledTargets: 0,
+				},
+			},
+			expState: "Ready",
+		},
+		"Pool state as Degraded": {
+			pqr: &PoolQueryResp{
+				Status: 0,
+				UUID:   "foo",
+				State:  "Ready",
```
Comment (tanabarr) [inline src/control/lib/control/pool_test.go:531]: I find this a little awkward overwriting the pool state based on whether any targets have been disabled, I realise we have already discussed this and accepted it as a solution but I would be more comfortable with adding another field to indicate we are calculating another state based on whether targets are disabled e.g. TargetsDisabled or simply compare TotalTargets and DisabledTargets when printing the QueryResponse in the pretty package. Then there is less possibility of problems later on due to overwriting the state returned from PoolQuery.

**Review by tanabarr — DISMISSED** [review]

File: src/control/lib/control/pool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/lib/control/pool.go:617]: This shouldn't be exposed as a new API function. Instead, it could be a method on the `PoolQueryResp` struct, e.g. `.UpdateState()`. You could then call the method in `mgmtSvc.PoolQuery()` and also in `daos.poolQueryCmd.Execute()` to avoid duplicated logic.

File: src/control/lib/control/pool.go [hunk 13]
```diff
@@ -429,6 +429,7 @@ type (
 	// PoolQueryResp contains the pool query response.
 	PoolQueryResp struct {
 		Status int32  `json:"status"`
+		State  string `json:"state"`
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/lib/control/pool.go:432]: I think that if you change the type to `system.PoolServiceState`, that type's stringer will automatically take care of the conversion to string for JSON output. As noted in another comment, it would probably be best to add a new `Degraded` state.

**Review by mjmac — CHANGES_REQUESTED** [review]
Overall looks ok, but I think there should be a few improvements before landing.

File: src/control/lib/control/pool.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/control/lib/control/pool.go]: Updated the code based on discussion, and it's been implemented

**Comment by ravalsam** [comment]
Test failures are not related to this PR and defects are all ready open.
https://daosio.atlassian.net/browse/DAOS-14015
https://daosio.atlassian.net/browse/DAOS-14085

File: src/control/lib/control/pool.go [hunk 14]
```diff
@@ -608,25 +608,37 @@ func PoolQuery(ctx context.Context, rpcClient UnaryInvoker, req *PoolQueryReq) (
...
+	return pqr, err
 }
 
-func UpdatePoolQueryState(pqr *PoolQueryResp) {
+// Update the pool state
+func (pqr *PoolQueryResp) UpdateState() error {
 	// Pool state is unknown as default, if TotalTargets is 0.
 	if pqr.TotalTargets == 0 {
-		pqr.State = "Unknown"
+		pqr.State = system.PoolServiceStateUnknown.String()
```
Comment (tanabarr) [inline src/control/lib/control/pool.go]: Not ideal having the state stored as a string, should really store the PoolServiceState itself, is this done because of difficulty with type conversion and JSON output compatibility?

File: src/control/system/pool.go [hunk 15]
```diff
@@ -24,6 +24,10 @@ const (
 	PoolServiceStateReady
 	// PoolServiceStateDestroying indicates that the pool service is being destroyed
 	PoolServiceStateDestroying
+	// PoolServiceStateDegraded indicates that the pool service is being Degraded
```
Comment (tanabarr) [inline src/control/system/pool.go:31]: Should probably offer a bit more description about what "degraded" means here, also "being degraded" doesn't make sense in this context.

File: src/control/lib/control/pool.go [hunk 16]
```diff
@@ -428,9 +428,9 @@ type (
 
 	// PoolQueryResp contains the pool query response.
 	PoolQueryResp struct {
-		Status int32  `json:"status"`
-		State  string `json:"state"`
-		UUID   string `json:"uuid"`
+		Status int32                   `json:"status"`
+		State  system.PoolServiceState `json:"state"`
```
Comment (tanabarr) [inline src/control/lib/control/pool.go:436]: Great thanks for fixing this, please ignore the earlier comment about it as I'm having to go through the commits individually because on the PR there was no way to view changes since my last review.

File: src/control/lib/control/pool.go [hunk 17]
```diff
@@ -1229,7 +1229,7 @@ func ListPools(ctx context.Context, rpcClient UnaryInvoker, req *ListPoolsReq) (
 			return nil, errors.New("pool query response uuid does not match request")
 		}
 
-		p.State = resp.State
+		p.State = resp.State.String()
```
Comment (tanabarr) [inline src/control/lib/control/pool.go:1238]: why does the state need to be held as a string here? is that inconsistent with PoolQueryResponse?

File: src/tests/ftest/control/dmg_pool_query_test.yaml [hunk 18]
```diff
@@ -27,8 +27,8 @@ ior:
 exp_vals:
   pool_status: 0
   pool_state: "Ready"
-  total_targets: 8
-  active_targets: 8
+  total_targets: 4
```
Comment (tanabarr) [inline src/tests/ftest/control/dmg_pool_query_test.yaml:30]: Rather than fix it here you should probably just merge with master to pick up the fix

File: src/proto/mgmt/pool.proto [hunk 19]
```diff
@@ -206,6 +206,14 @@ message PoolRebuildStatus {
 	uint64 records = 4;
 }
 
+enum PoolServiceState {
+	Creating = 0;   // pool service is being created
+	Ready = 1;      // pool service is ready to be used
+	Destroying = 2; // pool service is being destroyed
+	Degraded = 3 ;  // pool service is being Degraded
```
Comment (tanabarr) [inline src/proto/mgmt/pool.proto:213]: sp: "being Degraded" -> "degraded"

**Comment by tanabarr** [comment]
None of the comments are really blocking, please address them in the PR only if you have to re-push.

File: src/control/lib/control/pool.go (hunk 17 above)
Comment (ravalsam) [reply src/control/lib/control/pool.go:1238]: Well this belongs to ListPoolsResp which is an array of pools and state is string, and I kept it as it is.
https://github.com/daos-stack/daos/blob/bfc2ee0bd7eddac3109bb556ec2ed799e2f480da/src/proto/mgmt/pool.proto#L147C34-L147C34

File: src/control/lib/control/pool.go [hunk 20]
```diff
@@ -612,7 +612,37 @@ func PoolQuery(ctx context.Context, rpcClient UnaryInvoker, req *PoolQueryReq) (
...
+	err = pqr.UpdateState()
+	if err != nil {
+		return nil, err
+	}
+
+	return pqr, err
+}
+
+// Update the pool state
+func (pqr *PoolQueryResp) UpdateState() error {
```
Comment (kjacque) [inline src/control/lib/control/pool.go:629]: Is there a ticket to address this on the engine side by adding a degraded state at the engine level? It would be good to add a TODO referencing that ticket number here. Whoever works on that code will need to know this can be removed.

File: src/control/lib/control/pool.go [hunk 21]
```diff
@@ -1056,7 +1086,7 @@ type (
 		Usage []*PoolTierUsage `json:"usage"`
 
 		// PoolRebuildStatus contains detailed information about the pool rebuild process.
```
Comment (kjacque) [inline src/control/lib/control/pool.go:1088]: Comment still needs to be fixed
```suggestion
		// RebuildState contains detailed information about the pool rebuild process.
```

File: src/control/lib/control/pool.go [hunk 22]
```diff
@@ -612,7 +612,37 @@ func PoolQuery(ctx context.Context, rpcClient UnaryInvoker, req *PoolQueryReq) (
...
+
+	err = pqr.UpdateState()
+	if err != nil {
+		return nil, err
+	}
+
+	return pqr, err
+}
+
+// Update the pool state
```
Comment (kjacque) [inline src/control/lib/control/pool.go:628]: Nit - comment should begin with `UpdateState` for the sake of autogenerated Go documentation

**Review by mjmac — APPROVED** [review]

**Comment by ravalsam** [comment]
@daos-stack/daos-gatekeeper This PR is ready to merge
